### PR TITLE
Use SQLite for persistence (and Rust for handling the SQL)

### PR DESCRIPTION
**Pull request by arvid220u**
Refactor the daemon to save all data in SQLite, using Diesel in Rust to get type safety.

Benefits:
1. ACID data. Yay for reliability :)
2. More modular daemon structure

### CODE REVIEW

File: daemon-rs/transmitter/transmitter.cc [hunk 1]
```diff
@@ -105,7 +121,7 @@ auto Transmitter::retrieve() -> void {
   }
 
   setup_registration_caching();
-  auto& client = *cached_pir_client;
+  auto& client = **cached_pir_client;
```
Comment (sualehasif) [inline daemon-rs/transmitter/transmitter.cc:124]: @sualehasif fix this with the proper type safe methods

File: daemon-rs/transmitter/transmitter.cc [hunk 2]
```diff
@@ -131,7 +147,8 @@ auto Transmitter::retrieve() -> void {
   vector<db::Address> receive_addresses;
   vector<bool> receive_addresses_is_dummy;
   for (auto i = 0;
-       i < std::min(RECEIVE_FRIENDS_PER_ROUND, priority_receive_friends.size());
+       i < std::min(RECEIVE_FRIENDS_PER_ROUND,
+                    static_cast<int>(priority_receive_friends.size()));
```
Comment (sualehasif) [inline daemon-rs/transmitter/transmitter.cc:151]: Casting bug?? Narrowing conversinon

File: daemon-rs/transmitter/transmitter.cc [hunk 3]
```diff
@@ -215,33 +245,41 @@ auto Transmitter::retrieve() -> void {
...
-                                .chunks_start_sequence_number =
-                                    chunk.chunks_start_sequence_number,
-                                .num_chunks = chunk.num_chunks(),
-                                .s = chunk.msg()});
+        // TODO: we probably don't want to cast to int32 here... let's use
+        // int64s everywhere
+        auto received_full_message = G.db->receive_chunk(
+            (db::IncomingChunkFragment){
+                .from_friend = f.uid,
+                .sequence_number = static_cast<int>(chunk.sequence_number()),
```
Comment (sualehasif) [inline daemon-rs/transmitter/transmitter.cc:265]: narrowing conversion.

Comment (sualehasif) [reply daemon-rs/transmitter/transmitter.cc:265]: ig, it is ok thought.

**Review by sualehasif — COMMENTED** [review]
Seems gg for now :)

File: daemon-rs/transmitter/transmitter.cc (hunk 2 above)
Comment (arvid220u) [reply daemon-rs/transmitter/transmitter.cc]: can be at most 2. so not a problem

File: daemon-rs/transmitter/transmitter.cc (hunk 3 above)
Comment (arvid220u) [reply daemon-rs/transmitter/transmitter.cc]: ok for now, but we should think about whether we need to move rom int32s to int64s

File: daemon-rs/transmitter/transmitter.cc (hunk 1 above)
Comment (arvid220u) [reply daemon-rs/transmitter/transmitter.cc]: seems fine to me, but leaving this conversation open if you want to change it

**Pull request merged**